plugin/src/adapter100/kotlin/org/jlleitschuh/gradle/ktlint/worker/KtLintInvocation100.kt

```diff
@@ -34,11 +39,62 @@ class KtLintInvocation100(
             return KtLintInvocation100(engine)
         }
 
-        private fun loadRuleSetsFromClasspathWithRuleSetProviderV3(): Set<RuleProvider> {
-            return ServiceLoader
+        private fun loadRuleSetsFromClasspathWithRuleSetProviderV3(maxRuleVersion: String?): Set<RuleProvider> {
+            val allRuleProviders = ServiceLoader
                 .load(RuleSetProviderV3::class.java)
                 .flatMap { it.getRuleProviders() }
                 .toSet()
+
+            return if (maxRuleVersion != null) {
+                filterRulesByVersion(allRuleProviders, maxRuleVersion)
+            } else {
+                allRuleProviders
+            }
+        }
+
+        private fun filterRulesByVersion(ruleProviders: Set<RuleProvider>, maxVersion: String): Set<RuleProvider> {
+            return ruleProviders.filter { ruleProvider ->
+                isRuleCompatibleWithVersion(ruleProvider, maxVersion)
+            }.toSet()
+        }
+
+        private fun isRuleCompatibleWithVersion(ruleProvider: RuleProvider, maxVersion: String): Boolean {
+            // Use reflection to check for @SinceKtlint annotation
+            val ruleClass = ruleProvider.createNewRuleInstance()::class.java
+            val sinceAnnotation = ruleClass.getAnnotation(SinceKtlint::class.java)
+
+            return if (sinceAnnotation != null) {
+                isVersionCompatible(sinceAnnotation.version, maxVersion)
+            } else {
+                // If no annotation, assume it's from an older ktlint version (before annotations were included
+                // at runtime) and include it for backward compatibility
+                true
+            }
+        }
```

Comment:
This filtering isn't a feature built into ktlint itself as a library? I guess we are doing the rule loading manually, so I suppose this makes sense.